### PR TITLE
Add Concat Map operator

### PR DESCRIPTION
**Pull request by sergdort**
Implemented `concatMap` operator for the Observable regarding #1093 

### Description of change

* Refactor `MergeLimitedSink` to make interface similar to `MergeSink`
* Add `MapConcatSink `
* Add `ConcatMap` producer
* Add `concatMap` method to the extension to the `ObservableType`

### CODE REVIEW

File: RxSwift/Observables/Implementations/Merge.swift [hunk 1]
```diff
@@ -56,12 +56,32 @@ fileprivate final class MergeLimitedSinkIter<S: ObservableConvertibleType, O: Ob
...
+}
+
+fileprivate final class MergeLimitedBasicSink<S: ObservableConvertibleType, O: ObserverType>: MergeLimitedSink<S, S, O> where O.E == S.E {
+    
+    override func performMap(_ element: S) throws -> S {
+        return element
+    }
+}
+
+fileprivate class MergeLimitedSink<SourceType, S: ObservableConvertibleType, O: ObserverType>
```
Comment (sergdort) [inline RxSwift/Observables/Implementations/Merge.swift]: Not sure if it good idea to extend `MergeLimitedSink ` to be similar to `MergeSink`. 
In theory we could either subclass from `MergeSink` or create `ConcatMapSink` without subclassing from `MergeLimitedSink`

Comment (sergdort) [inline RxSwift/Observables/Implementations/Merge.swift]: Then we won't need to have  `MergeLimitedBasicSink`

File: RxSwift/Observables/Implementations/Merge.swift [hunk 2]
```diff
@@ -408,6 +440,24 @@ final class FlatMapFirst<SourceType, S: ObservableConvertibleType>: Producer<S.E
...
+    private let _source: Observable<SourceType>
+    private let _selector: Selector
+    
+    init(source: Observable<SourceType>, selector: @escaping Selector) {
+        _source = source
+        _selector = selector
+    }
+    
+    override func run<O: ObserverType>(_ observer: O, cancel: Cancelable) -> (sink: Disposable, subscription: Disposable) where O.E == S.E {
+        let sink = MapConcatSink<SourceType, S, O>(selector: _selector, observer: observer, cancel: cancel)
```
Comment (sergdort) [inline RxSwift/Observables/Implementations/Merge.swift]: This is probable should be `ConcatMapSink`?

File: RxSwift/Observables/Implementations/Merge.swift (hunk 1 above)
Comment (kzaher) [inline RxSwift/Observables/Implementations/Merge.swift]: Hi @sergdort ,

yeah, I think it's ok to make `MergeLimitedSink` a bit more generic and then reuse the logic.

I've reordered project structure a bit in `develop`. Can you please add unit tests for this and target `develop` branch?

We can reuse and modify `flatMap` unit tests a bit.

File: RxSwift/Observables/Implementations/Merge.swift (hunk 2 above)
Comment (kzaher) [inline RxSwift/Observables/Implementations/Merge.swift]: Yep

**Comment by sergdort** [comment]
Hi, @kzaher sorry for the delay, I was on holidays =)

I like that you moved public interface into implementation although I've noticed that sinks comes before producers, is it intentional if so I would like to know why :)
Because I feel like it much more natural to have classes and method in the order  they are used, sure this is kind of opinionated but ¯\_(ツ)_/¯ (Uncle Bob will be happy) 😆

**Comment by sergdort** [comment]
I've also found some tests from [RxJS](https://github.com/Reactive-Extensions/RxJS/blob/master/tests/observable/concatmap.js#L145) I think we can reuse them

**Comment by kzaher** [comment]
> I like that you moved public interface into implementation although I've noticed that sinks comes before producers, is it intentional if so I would like to know why :)

I think it was because all sequence declarations were the same, so they were practically irrelevant. Maybe there was some other rationale, but can't remember now.

> Because I feel like it much more natural to have classes and method in the order they are used, sure this is kind of opinionated but ¯_(ツ)_/¯ (Uncle Bob will be happy) 😆

I think it makes sense, we could move them in front of `Sink`s.

**Comment by sergdort** [comment]
Seems like CI is passing :)

**Comment by kzaher** [comment]
Hi @sergdort ,

did you maybe rebase?

**Comment by sergdort** [comment]
I'm little bit confused :) Why do I need to rebase?

**Comment by kzaher** [comment]
Hi @sergdort,

because I can't do a CR or merge this PR because of conflicts.

screen shot 2017-05-17 at 00 34 33

**Comment by sergdort** [comment]
Done

**Pull request merged**